File: packages/web/src/utils/jsonc.test.ts
import { describe, it, expect } from 'vitest';
import { stripJSONC } from './jsonc.js';

describe('stripJSONC utility', () => {
    it('should keep plain JSON unchanged and preserve its length', () => {
        const input = '{"a": 1, "b": "hello"}';
        const output = stripJSONC(input);
        expect(output).toBe(input);
        expect(output.length).toBe(input.length);
    });

    it('should strip single-line comments and preserve length', () => {
        const input = `{\n  // this is a comment\n  "a": 1\n}`;
        const output = stripJSONC(input);
        // Comments should be replaced with spaces, keeping newlines
        expect(output).toBe(`{\n                      \n  "a": 1\n}`);
        expect(output.length).toBe(input.length);
        expect(JSON.parse(output)).toEqual({ a: 1 });
    });

    it('should strip inline single-line comments', () => {
        const input = `{\n  "a": 1 // some comment\n}`;
        const output = stripJSONC(input);
        expect(output).toBe(`{\n  "a": 1                \n}`);
        expect(output.length).toBe(input.length);
        expect(JSON.parse(output)).toEqual({ a: 1 });
    });

    it('should strip multi-line block comments and preserve line endings', () => {
        const input = `{\n  /*\n  block\n  comment\n  */\n  "a": 1\n}`;
        const output = stripJSONC(input);
        expect(output).toBe(`{\n    \n       \n         \n    \n  "a": 1\n}`);
        expect(output.length).toBe(input.length);
        expect(JSON.parse(output)).toEqual({ a: 1 });
    });

    it('should ignore comment characters inside string values', () => {
        const input = '{"url": "http://example.com/path", "text": "/* not comment */", "slash": "// not comment"}';
        const output = stripJSONC(input);
        expect(output).toBe(input);
        expect(output.length).toBe(input.length);
        expect(JSON.parse(output)).toEqual({
            url: "http://example.com/path",
            text: "/* not comment */",
            slash: "// not comment"
        });
    });

    it('should handle nested escaped quotes and other escape sequences', () => {
        const input = '{"text": "hello \\"world // comment? no\\"!"}';
        const output = stripJSONC(input);
        expect(output).toBe(input);
        expect(output.length).toBe(input.length);
        expect(JSON.parse(output)).toEqual({
            text: 'hello "world // comment? no"!'
        });
    });

    it('should handle empty input gracefully', () => {
        const input = '';
        const output = stripJSONC(input);
        expect(output).toBe('');
    });

    it('should handle trailing potential comment slash', () => {
        const input = '{"a": 1}/';
        const output = stripJSONC(input);
        expect(output).toBe('{"a": 1}/');
    });

    it('should handle multiple asterisks in block comments', () => {
        const input = '{"a": 1} /***/';
        const output = stripJSONC(input);
        expect(output).toBe('{"a": 1}      ');
        expect(output.length).toBe(input.length);
    });

    it('should handle empty block comments', () => {
        const input = '{"a": 1} /**/';
        const output = stripJSONC(input);
        expect(output).toBe('{"a": 1}     ');
        expect(output.length).toBe(input.length);
    });
});

File: packages/container/jsonc_test.go
package main

import (
	"encoding/json"
	"testing"

	"swazz-engine/internal/swagger"

	"github.com/stretchr/testify/assert"
)

func TestStripJSONC(t *testing.T) {
	tests := []struct {
		name        string
		input       string
		expectedVal map[string]any
		isValid     bool
	}{
		{
			name:  "plain json unchanged",
			input: `{"a": 1, "b": "hello"}`,
			expectedVal: map[string]any{
				"a": float64(1),
				"b": "hello",
			},
			isValid: true,
		},
		{
			name: "single line comment on its own line",
			input: `{
				// this is a comment
				"a": 1
			}`,
			expectedVal: map[string]any{
				"a": float64(1),
			},
			isValid: true,
		},
		{
			name: "inline line comment after value",
			input: `{
				"a": 1, // comment here
				"b": 2
			}`,
			expectedVal: map[string]any{
				"a": float64(1),
				"b": float64(2),
			},
			isValid: true,
		},
		{
			name: "multi-line block comment",
			input: `{
				/* 
				comment block
				spanning lines
				*/
				"a": 1
			}`,
			expectedVal: map[string]any{
				"a": float64(1),
			},
			isValid: true,
		},
		{
			name:  "comment inside a string must not be stripped",
			input: `{"url": "http://example.com/path", "text": "/* not a comment */", "slash": "// not comment"}`,
			expectedVal: map[string]any{
				"url":   "http://example.com/path",
				"text":  "/* not a comment */",
				"slash": "// not comment",
			},
			isValid: true,
		},
		{
			name:  "nested escaped quotes",
			input: `{"text": "hello \"world // comment? no\"!"}`,
			expectedVal: map[string]any{
				"text": `hello "world // comment? no"!`,
			},
			isValid: true,
		},
		{
			name:    "empty input",
			input:   ``,
			isValid: false,
		},
		{
			name:    "trailing single slash",
			input:   `{"a": 1}/`,
			isValid: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			stripped := swagger.StripJSONC([]byte(tt.input))

			// Verify that line numbers/byte offsets remain identical
			assert.Equal(t, len(tt.input), len(stripped), "Length of stripped content must match input to preserve byte offsets")

			if tt.isValid {
				var v map[string]any
				err := json.Unmarshal(stripped, &v)
				assert.NoError(t, err, "Stripped JSON must be valid parsing result")
				assert.Equal(t, tt.expectedVal, v)
			} else if tt.input != "" {
				// For invalid JSON but not empty, parsing should fail after stripping
				var v map[string]any
				err := json.Unmarshal(stripped, &v)
				assert.Error(t, err)
			}
		})
	}
}
